File: Scala/algorithmic_toolbox/src/main/scala-2.12/divideandconquer/ClosestPairOfPoints.scala
package divideandconquer

object ClosestPairOfPoints {
  import Point.distance
  private val BruteForceSize: Int = 3

  final case class Point(x: Double, y: Double)

  object Point {
    def distance(p: Point, q: Point): Double = math.hypot(p.x - q.x, p.y - q.y)
  }

  private def convertToDoubleList(line: String): List[Double] = line.split(" ").map(_.toDouble).toList

  private def calcSmallestPairwiseDistance(points: Vector[Point],
                                           minDistance: Double,
                                           nrPointsCompareWith: Int): Double = {
    def calcDistances(currentMinDistance: Double, pointAndIndex: (Point, Int)): Double = {
      val (p, ix): (Point, Int) = pointAndIndex
      points
        .slice(ix + 1, ix + nrPointsCompareWith + 1)
        .foldLeft(currentMinDistance) { case (acc, q) =>
          val dist: Double = distance(p, q)
          if (dist < acc) dist else acc
      }
    }
    points.zipWithIndex.foldLeft(minDistance)(calcDistances)
  }

  private def findPointsInStripe(first: Vector[Point],
                                 second: Vector[Point],
                                 median: Double,
                                 delta: Double): Vector[Point] = {
    def inStripe(p: Point): Boolean = math.abs(p.x - median) <= delta
    first.filter(inStripe) ++ second.filter(inStripe)
  }

  private def calcMinimumDistanceInStripe(first: Vector[Point],
                                          second: Vector[Point],
                                          median: Double,
                                          delta: Double): Double = {
    val stripe: Vector[Point] = findPointsInStripe(first, second, median, delta)
    calcSmallestPairwiseDistance(stripe.sortBy(_.y), delta, 7)
  }

  private def findClosestPoints(sortedPoints: Vector[Point], n: Int): Double = {
    if (n <= BruteForceSize) calcSmallestPairwiseDistance(sortedPoints, Double.MaxValue, BruteForceSize - 1)
    else {
      val middleIx: Int = n / 2
      val medianX: Double = sortedPoints(middleIx).x
      val (first, second): (Vector[Point], Vector[Point]) = sortedPoints.splitAt(middleIx)
      val delta1: Double = findClosestPoints(first, middleIx)
      val delta2: Double = findClosestPoints(second, n - middleIx)
      val delta: Double = math.min(delta1, delta2)
      if (math.abs(delta) < 1e-14) 0.0
      else calcMinimumDistanceInStripe(first, second, medianX, delta)
    }
  }

  def distanceOfClosestPairOfPoints(points: Vector[Point], n: Int): Double = {
    val sortedPoints: Vector[Point] = points.sortBy(_.x)
    findClosestPoints(sortedPoints, n)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val points: Vector[Point] = (for { _ <- 0 until n } yield {
      val List(x, y): List[Double] = convertToDoubleList(reader.next())
      Point(x, y)
    }).toVector
    val result: Double = distanceOfClosestPairOfPoints(points, n)
    println(result)
  }
}

File: OCaml/algorithmic_toolbox/divideandconquer/closest_pair_of_points.ml
open Batteries

type point = { x : float; y : float }

let brute_force_size = 3


let convert_to_floatlist line = List.map float_of_string Str.(line |> split (regexp " "))


let distance { x = x'; y = y'} { x = x''; y = y''} = hypot (x' -. x'') (y' -. y'')


let read_points n =
    let rec read_point n acc = match n with
        | 0 -> List.rev acc
        | _ -> read_point (n - 1) ((Scanf.sscanf (read_line ()) "%f %f" (fun x y -> { x; y })) :: acc)
    in read_point n []


let calc_smallest_pairwise_distance points min_distance ~nr_points_compare_with =
    let calc_distances current_min_distance (p, ix) =
        let slice = List.take nr_points_compare_with (List.drop (ix + 1) points) in
        let update_min_distance acc q =
            let dist = distance p q
            in if dist < acc then dist else acc
        in List.fold_left update_min_distance current_min_distance slice in
    let zipped_with_index = List.mapi (fun ix elem -> (elem, ix)) points
    in List.fold_left calc_distances min_distance zipped_with_index


let calc_minimum_distance_in_stripe first second median delta =
    let in_stripe { x } = abs_float (x -. median) <= delta in
    let cmp_by_y { y=y' } { y=y'' } = compare y' y'' in
    let stripe = (List.filter in_stripe first) @ (List.filter in_stripe second)
    in calc_smallest_pairwise_distance (List.sort cmp_by_y stripe) delta ~nr_points_compare_with:7


let rec find_closest_points sorted_points n =
    if n <= brute_force_size then
        calc_smallest_pairwise_distance sorted_points max_float ~nr_points_compare_with:(brute_force_size - 1)
    else
        let middle_ix = n / 2 in
        let median_x = (List.at sorted_points middle_ix).x in
        let first, second = List.split_at middle_ix sorted_points in
        let delta1 = find_closest_points first middle_ix in
        let delta2 = find_closest_points second (n - middle_ix) in
        let delta = min delta1 delta2 in
        if (abs_float delta) < 1e-14 then 0.0
        else calc_minimum_distance_in_stripe first second median_x delta


let find_closest_pair_of_points points n =
    let cmp_by_x { x=x' } { x=x'' } = compare x' x'' in
    let sorted_points = List.sort cmp_by_x points
    in find_closest_points sorted_points n


let () =
    let n = read_int() in
    let points = read_points n in
    let dist = find_closest_pair_of_points points n in
    print_float dist;
    print_newline()
